tfhe/src/integer/server_key/radix_parallel/sub.rs

```diff
@@ -402,6 +413,329 @@ impl ServerKey {
+        // compute either the overflow flag or the borrow flag as the user request
+        // but as the overflow flag is not needed in the code base, we simply only
+        // compute the borrow flag if requested.
+        //
+        // This is why the inputs are RadixCiphertext rather than &[Ciphertext]
+
+        let lhs = &mut lhs.blocks;
+        let rhs = &rhs.blocks;
+
+        assert_eq!(
+            lhs.len(),
+            rhs.len(),
+            "Left hand side must must have a number of blocks equal \
+            to the number of blocks of the right hand side: lhs {} blocks, rhs {} blocks",
+            lhs.len(),
+            rhs.len()
+        );
+
+        if lhs.is_empty() {
+            // Then both are empty
+            if compute_overflow {
+                return Some(self.create_trivial_boolean_block(false));
+            }
+            return None;
+        }
+
+        for (lhs_b, rhs_b) in lhs.iter_mut().zip(rhs.iter()) {
+            self.key.unchecked_sub_assign(lhs_b, rhs_b);
+        }
+        if let Some(borrow) = input_borrow {
+            self.key.unchecked_sub_assign(&mut lhs[0], &borrow.0);
+        }
+
+        // First step
+        let (shifted_blocks, mut block_states) =
+            self.compute_shifted_blocks_and_block_borrow_states(lhs);
+
+        // The propagation state of the last block will be used to determine
+        // if overflow occurs (i.e is there an output borrow)
+        let mut overflow_block = block_states.pop().unwrap();
+
+        let block_modulus = self.message_modulus().0 * self.carry_modulus().0;
+        let num_bits_in_block = block_modulus.ilog2();
+
+        // Just in case we compare with max noise level, but it should always be num_bits_in_blocks
+        // with the parameters we provide
+        let grouping_size = (num_bits_in_block as usize).min(self.key.max_noise_level.get());
+
+        // Second step
+        let (mut prepared_blocks, resolved_borrows) = {
+            let (propagation_simulators, resolved_borrows) = self
+                .compute_propagation_simulators_and_groups_carries(grouping_size, &block_states);
+
+            let mut prepared_blocks = shifted_blocks;
+            prepared_blocks
+                .iter_mut()
+                .zip(propagation_simulators.iter())
+                .for_each(|(block, simulator)| {
+                    // simulator may have either of these value
+                    // '2' if the block is borrowed from
+                    // '1' if the block will be borrowed from if the group it belongs to receive a
+                    //     borrow
+                    // '0' if the block will absorb any potential borrow
+                    //
+                    // What we do is we subtract this value from the block, as it's a borrow, not a
+                    // carry, and we add one, this means:
+                    //
+                    // '- (2 + 1) ==  -1' We remove one if the block was meant to receive a borrow
```

Comment:
wrong sign in the parens or parens not in the right place ? same for all three

Reply:
Ah right it should be

```
-2 + 1 = -1
-1 + 1 = 0
0 + =  1
```